### ISSUE TITLE
Admin password in crash.log

### ISSUE THREAD

**Issue by ddeboer** [zotonic/zotonic#1666]
The crash log contains an `{admin_password, PASSWORD}` tuple. Can we filter this out to improve security?

```
2017-04-20 11:48:51 =SUPERVISOR REPORT====
     Supervisor: {local,'z_site_sup$some_site'}
     Context:    shutdown_error
     Reason:     shutdown
     Offender:   [{pid,<0.9617.107>},{id,z_installer},{mfargs,{z_installer,start_link,[[{title,<<"Some Site">>},{skeleton,empty},{seo,[{noindex,false}]},{redirect,true},{pagelen,5},{install_modules,[mod_base,mod_acl,mod_authentication,mod_content_groups,mod_acl_user_groups,mod_acl_adminonly,mod_admin,mod_oauth,mod_search,mod_atom_feed,mod_translation,mod_signal,mod_logging,mod_mqtt,mod_menu,mod_editor_tinymce,mod_seo,mod_seo_google,mod_seo_sitemap,mod_oembed,mod_admin_category,mod_admin_config,mod_admin_identity,mod_admin_modules,mod_admin_predicate,mod_filestore,mod_l10n,mod_geo,mod_artwork,mod_bootstrap,mod_video,mod_video_embed,mod_ginger_base,mod_ginger_foundation,mod_ginger_admin,mod_ginger_edit,mod_ginger_auth,mod_ginger_search,mod_ginger_activity,mod_import_anymeta_dispatch,mod_alternative_uris]},{hostname,"www.somesite.com"},{host,some_site},{enabled,true},{depcache_memory_max,100},{dbschema,"public"},{dbdatabase,"mediamatic"},{admin_password,"SECRET_PASSWORD"}]]}},{restart_type,permanent},{shutdown,1},{child_type,worker}]
```

**Comment by mmzeeman** [comment]
Mind that the admin password is readable in the config file too.

**Comment by mmzeeman** [comment]
Idea. 

Maybe we can place the admin password in a little dedicated process which we flag as sensitive. Than it will no longer appear in crash logs and you will not be able to fetch it with debug tools.

**Comment by mworrell** [comment]
We want to change the site supervisor anyway. This for better handling of the startup process and incoming requests for sites that are not yet started or restarting.

The new supervisor could then be started regardless if the site is enabled or not (and load/inspect/watch the config files for any changes).

Besides this we should store all config settings in an ets table for faster retrieval.

**Comment by crownedgrouse** [comment]
Password should not be stored but only its hash.
Password given in login has to be hashed and compared to this hash.

**Comment by mmzeeman** [comment]
Thanks for bringing this back into our attention. 👍

This is the password of the first user in the system "admin", which is in the config of the site. All other users's password are stored as bcrypt hash values.

It might indeed be an idea to put the password of admin in hashed form in the site's config file. With the help of a little command line utility it should be possible to calculate this hash, and put it in the site's config.

**Comment by crownedgrouse** [comment]
A first step would to immediately hash the password at start and then keep it like that on runtime, so that crash log don't contain useful info.

**Comment by mmzeeman** [comment]
That is indeed easier. I forgot we also accept JSON and YAML site config files.

**Comment by mmzeeman** [comment]
Are there other configuration values which should not end up in logs? The database password maybe?

I guess this is a feature which is really needed when you run Zotonic in a container. When the logs are stored in the same location as the config files it is not really an issue.

**Comment by crownedgrouse** [comment]
Except database password is probably to be kept unchanged in order to connect to database. Not the same for admin password which is on a server side.

**Comment by mmzeeman** [comment]
This is a `0.x` issue. In `1.0` the config of the site is loaded into `application` environment variables. Because they are not kept in the supervisor anymore, they do not end up in logs with `1.0`. 

I've looked at how other erlang projects which have sensitive data in their config store these values. RabbitMQ allows encrypted config settings. https://www.rabbitmq.com/configure.html#configuration-encryption.

**Comment by crownedgrouse** [comment]
yes but as said in this documentation, it is not really a security as there is  a need of a clear passphrase. this only hide the real password, and worse do not secure it. I think we should also different password format plain or hashed. For second option, password have to be kept by admin elsewhere and cannot be guess on runtime. For instance :

```
"password": "p455w0rd"
"password": "md5:a4fd8e6fa9fbf9a6f2c99e7b70aa9ef2"
"password": "sha1:b4341ce88a4943631b9573d9e0e5b28991de945d"
```
I think it is a better option to allow an easy password change only in command line (proving a ssh based authentication probably), and store it in specific file, not in config. So that different access right can be set. 
chown root:zotonic  and  chmod 740

**Comment by crownedgrouse** [comment]
for first option, store only hashed plain text password in runtime code, with a warning log encouraging to use another way.

**Comment by mmzeeman** [comment]
Yes we put this on the list for 1.1. 

FYI it is possible to disable the admin user by unpublishing it’s resource. Then only users with bcrypt hashes are used.

**Comment by crownedgrouse** [comment]
See also Radius authentication.
https://github.com/travelping/eradius  for instance
